Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,142 @@
+import enum
+
+from types import TracebackType
+from typing import final, Any, Dict, Optional, Type
+
+from . import events
+from . import tasks
+
+
+__all__ = (
+    "Timeout",
+    "timeout",
+    "timeout_at",
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    EXPIRING = "expiring"
+    EXPIRED = "expired"
+    EXITED = "exited"
+
+
+@final
+class Timeout:
+
+    def __init__(self, deadline: Optional[float]) -> None:
+        self._state = _State.CREATED
+
+        self._timeout_handler: Optional[events.TimerHandle] = None
+        self._task: Optional[tasks.Task[Any]] = None
+        self._deadline = deadline
+
+    @property
+    def deadline(self) -> Optional[float]:
+        return self._deadline
+
+    def reschedule(self, deadline: Optional[float]) -> None:
+        assert self._state != _State.CREATED
+        if self._state != _State.ENTERED:
+            raise RuntimeError(
+                f"Cannot change state of {self._state} CancelScope",
+            )
+
+        self._deadline = deadline
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+
+        if deadline is None:
+            self._timeout_handler = None
+        else:
+            loop = events.get_running_loop()
+            self._timeout_handler = loop.call_at(
+                deadline,
+                self._on_timeout,
+            )
+
+    def expired(self) -> bool:
+        """Is timeout expired during execution?"""
+        return self._state in (_State.EXPIRING, _State.EXPIRED)
+
+    def __repr__(self) -> str:
+        info = [str(self._state)]
+        if self._state is _State.ENTERED and self._deadline is not None:
+            info.append(f"deadline={self._deadline}")
+        cls_name = self.__class__.__name__
+        return f"<{cls_name} at {id(self):#x}, {' '.join(info)}>"
+
+    async def __aenter__(self) -> "Timeout":
```

Comment:
Is everyone now on board with this being an async context manager?

Reply:
I plan on maintaining non-async versions in Quattro, so following the comments about consistency and ease of teaching I'm totally fine with it.

Reply:
I wonder if we should add dummy `__enter__` here (and to TaskGroup!) that raises an error telling the user to use `async with`. I know I've been puzzled a few times already when I forgot that `async` when playing around with TaskGroup.

Reply:
Not sure. We removed `__enter__`/`__exit__` from locks recently. The default message for the missing method is good enough for me; we can improve it probably. Say, suggest `async with` if `__enter__` is missing but `__aenter__` is present.

Reply:
Good point, this should be fixed (if it really matters) inside `async with` and `with`, not in each context manager.